tests/queries/0_stateless/01576_alias_column_rewrite.sql

```diff
@@ -0,0 +1,112 @@
+INSERT INTO test_table(timestamp, value) SELECT toDateTime('2020-01-03 12:00:00'), 1 FROM numbers(10);
+
+set optimize_alias_column_prediction = 1;
+SELECT 'test-partition-prune';
+
+SELECT COUNT() = 10 FROM test_table WHERE day = '2020-01-01' SETTINGS max_rows_to_read = 10;
+SELECT t = '2020-01-03' FROM (SELECT day AS t FROM test_table WHERE t = '2020-01-03' GROUP BY t SETTINGS max_rows_to_read = 10);
+SELECT COUNT() = 10 FROM test_table WHERE day = '2020-01-01' UNION ALL SELECT 1 FROM numbers(1) SETTINGS max_rows_to_read = 11;
+SELECT  COUNT() = 0 FROM (SELECT  toDate('2019-01-01') AS  day, day AS t   FROM test_table PREWHERE t = '2020-01-03'  WHERE t  = '2020-01-03' GROUP BY t );
+
+
+
+SELECT 'test-join';
+SELECT day = '2020-01-03'
+FROM
+(
+ SELECT toDate('2020-01-03') AS day
+ FROM numbers(1)
+) AS a
+INNER JOIN
+(
+ SELECT day
+ FROM test_table
+ WHERE day = '2020-01-03'
+ GROUP BY day
+) AS b ON a.day = b.day SETTINGS max_rows_to_read = 11;
+
+SELECT day = '2020-01-01'
+FROM
+(
+ SELECT day
+ FROM test_table
+ WHERE day = '2020-01-01'
+ GROUP BY day
+) AS a
+INNER JOIN
+(
+ SELECT toDate('2020-01-01') AS day
+ FROM numbers(1)
+) AS b ON a.day = b.day SETTINGS max_rows_to_read = 11;
+
+
+SELECT 'alias2alias';
+SELECT COUNT() = 10 FROM test_table WHERE day1 = '2020-01-02' SETTINGS max_rows_to_read = 10;
+SELECT t = '2020-01-03' FROM (SELECT day1 AS t FROM test_table WHERE t = '2020-01-03' GROUP BY t SETTINGS max_rows_to_read = 10);
+SELECT t = '2020-01-03' FROM (SELECT day2 AS t FROM test_table WHERE t = '2020-01-03' GROUP BY t SETTINGS max_rows_to_read = 10);
+SELECT COUNT() = 10 FROM test_table WHERE day1 = '2020-01-03' UNION ALL SELECT 1 FROM numbers(1) SETTINGS max_rows_to_read = 11;
+SELECT  COUNT() = 0 FROM (SELECT  toDate('2019-01-01') AS  day1, day1 AS t   FROM test_table PREWHERE t = '2020-01-03'  WHERE t  = '2020-01-03' GROUP BY t );
+SELECT day1 = '2020-01-04' FROM test_table PREWHERE day1 = '2020-01-04'  WHERE day1 = '2020-01-04' GROUP BY day1 SETTINGS max_rows_to_read = 10;
+
+
+ALTER TABLE test_table add column array Array(UInt8) default [1, 2, 3];
+ALTER TABLE test_table add column struct.key Array(UInt8) default [2, 4, 6], add column struct.value Array(UInt8) alias array;
+
+SELECT 'array-join';
+set max_rows_to_read = 10;
+SELECT count() == 10 FROM test_table WHERE day = '2020-01-01';
+SELECT sum(struct.key) == 30, sum(struct.value) == 30 FROM (SELECT struct.key, struct.value FROM test_table array join struct WHERE day = '2020-01-01');
+
+SELECT 'lambda';
+-- lambda parameters in filter should not be rewrite
+SELECT count() == 10 FROM test_table WHERE  arrayMap((day) -> day + 1, [1,2,3]) [1] = 2 AND day = '2020-01-03';
+
+set max_rows_to_read = 0;
+-- how to test it? currently just check logs, eg: 00940_order_by_read_in_order
+SELECT 'optimize_read_in_order';
+SET optimize_read_in_order = 1;
+SELECT day AS s FROM test_table ORDER BY s LIMIT 1;
```

Comment:
```suggestion
SELECT day AS s FROM test_table ORDER BY s LIMIT 1 SETTINGS optimize_read_in_order=1;
```

It is better to use SETTINGS clause (I guess you can safely do this after #16619 ), since this will not affect any subsequent query